mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java

```diff
@@ -209,14 +209,29 @@ public Receive createReceive() {
 
     private void onAddNewJobArtifactsToCacheRequest(AddNewJobArtifactsToCacheRequest req) {
         try {
-            mantisJobStore.addNewJobArtifactsToCache(req.getClusterID(), req.getArtifacts());
-            jobArtifactsToCache.addAll(req.artifacts);
+            List<ArtifactID> newArtifacts = req.artifacts
+                .stream()
+                .filter(artifactID -> !jobArtifactsToCache.contains(artifactID))
+                .collect(Collectors.toList());
+
+            if (!newArtifacts.isEmpty()) {
+                log.info("Storing and caching new artifacts: {}", newArtifacts);
+
+                mantisJobStore.addNewJobArtifactsToCache(req.getClusterID(), newArtifacts);
+                jobArtifactsToCache.addAll(newArtifacts);
+                refreshTaskExecutorJobArtifactCache();
```

Comment:
Any potential impact on TE startup if the cache list is long or contains a few large artifacts?

Reply:
Today TEs can receive job requests while downloading job artifacts so the answer to your question is no. However this can be a problem in the future if we want to only accept requests _after_ all job artifacts have been downloaded.